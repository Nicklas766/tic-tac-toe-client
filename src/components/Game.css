.game {
    display: flex;
    flex-wrap: wrap;
    width: 50%;
    margin: auto;
}

.game-board, .game-info {
    width: 100%;
}

.game-info div {
    text-align: center;
    font-size: 2em;
}

.board-row {
    display: flex;
    flex-wrap: wrap;
    width: 100%;
}
  
.square {
    width: 33.33%;
    height: 100px;
    background: white;
    font-size: 2em;
}