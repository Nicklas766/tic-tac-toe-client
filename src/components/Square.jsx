import React from 'react';
import PropTypes from 'prop-types';

const Square = ({ onClick, value = '' }) => (
  <button type="submit" onClick={onClick} className="square">
    {value}
  </button>
);

Square.propTypes = {
  onClick: PropTypes.func.isRequired,
  value: PropTypes.string,
};

Square.defaultProps = {
  value: '',
};

export default Square;
