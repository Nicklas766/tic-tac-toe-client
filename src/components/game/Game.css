.game {
    display: flex;
    flex-wrap: wrap;
    width: 90%;
    margin: auto;
}

@media (min-width: 650px) {
    .game {
        width: 50%;
    }
}

.game-board, .game-info {
    width: 100%;
}

.game-info div {
    text-align: center;
    font-size: 2em;
}

.board-row {
    display: flex;
    flex-wrap: wrap;
    width: 100%;
}
  
.square {
    width: 33.33%;
    height: 100px;
    background: white;
    font-size: 2em;
}