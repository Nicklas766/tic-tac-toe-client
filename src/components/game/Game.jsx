import React, { useState } from 'react';
import {
  xIsNext, setSquare, squareCanBeSet, getWinner, allSquaresSet,
} from 'tic-tac-toe-stateless-engine';
import Board from './Board';
import Square from './Square';
import './Game.css';

const getGameStatus = (squares) => {
  const winner = getWinner(squares);

  if (winner || allSquaresSet(squares)) {
    return winner ? `${winner} won!` : 'It\'s a draw!';
  }

  return `Next player: ${xIsNext(squares) ? 'X' : 'O'}`;
};

const Game = () => {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [currIndex, setCurrIndex] = useState(0);
  const squares = history[currIndex];

  const handleSquareClick = (index) => {
    if (squareCanBeSet(squares, index)) {
      const newSquares = setSquare(squares, index);

      setHistory([...history.slice(0, currIndex + 1), newSquares]);
      setCurrIndex(currIndex + 1);
    }
  };

  const renderSquare = (index) => (
    <Square key={index} onClick={() => handleSquareClick(index)} value={squares[index]} />
  );

  const moves = history.map((square, index) => {
    const btnText = index ? `Go to move ${index}` : 'Go to game start';

    return (
      <li key={square}>
        <button type="submit" onClick={() => setCurrIndex(index)}>{btnText}</button>
      </li>
    );
  });


  return (
    <div className="game">
      <div className="game-board">
        <Board renderSquare={renderSquare} />
      </div>

      <div className="game-info">
        <div>{getGameStatus(squares)}</div>
        <ol>{moves}</ol>
      </div>
    </div>
  );
};

export default Game;
