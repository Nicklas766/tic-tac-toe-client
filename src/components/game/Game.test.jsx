import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Game from './Game';
import Square from './Square';

configure({ adapter: new Adapter() });

describe('Snapshot test', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Game />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('\n Make sure button text (X, O) changes on squares after clicking', () => {
  let wrap; let firstSquare; let
    lastSquare;

  beforeEach(() => {
    wrap = mount(<Game />);
    firstSquare = wrap.find(Square).first();
    lastSquare = wrap.find(Square).last();
  });

  it('renders empty string for first square', () => {
    expect(firstSquare.text()).toEqual('');
  });

  it('renders X for last square after clicking it', () => {
    lastSquare.simulate('click');

    expect(lastSquare.text()).toEqual('X');
  });

  it('renders same texts for first and last square after 0 clicks', () => {
    expect(firstSquare.text()).toEqual(lastSquare.text());
  });

  it('renders different texts for first and last square after 1 click', () => {
    firstSquare.simulate('click');

    expect(firstSquare.text()).not.toEqual(lastSquare.text());
  });
});

describe('\n Make sure status is correct during game and after game', () => {
  const getGameInfoText = (wrap) => wrap.find('.game').find('.game-info').childAt(0).text();
  let wrap; let
    squares;

  beforeEach(() => {
    wrap = mount(<Game />);
    squares = wrap.find(Square);
  });

  it('renders Next player X after 0 clicks', () => {
    expect(getGameInfoText(wrap)).toEqual('Next player: X');
  });

  it('renders Next player: O after 1 click', () => {
    squares.at(0).simulate('click');

    expect(getGameInfoText(wrap)).toEqual('Next player: O');
  });

  it('renders X wins! After X winning.', () => {
    squares.forEach((s) => s.simulate('click'));

    expect(getGameInfoText(wrap)).toEqual('X won!');
  });

  it('renders its a draw! After draw.', () => {
    [0, 4, 1, 2, 6, 3, 5, 7, 8].forEach((i) => squares.at(i).simulate('click'));

    expect(getGameInfoText(wrap)).toEqual("It's a draw!");
  });
});

describe('\n Make sure its possible to use history to backtrack', () => {
  const getGameInfoText = (wrap) => wrap.find('.game').find('.game-info').childAt(0).text();
  const getGoToGameStartButton = (wrap) => wrap.find('.game').find('.game-info').find('button').first();

  it('renders Next player X by using history after completing a game', () => {
    const wrap = mount(<Game />);
    const squares = wrap.find(Square);

    squares.forEach((s) => s.simulate('click'));

    expect(getGameInfoText(wrap)).toEqual('X won!');

    getGoToGameStartButton(wrap).simulate('click');

    expect(getGameInfoText(wrap)).toEqual('Next player: X');
  });
});
