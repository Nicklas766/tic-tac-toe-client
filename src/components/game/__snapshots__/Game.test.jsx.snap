// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test renders correctly 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
        <button
          className="square"
          onClick={[Function]}
          type="submit"
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div>
      Next player: X
    </div>
    <ol>
      <li>
        <button
          onClick={[Function]}
          type="submit"
        >
          Go to game start
        </button>
      </li>
    </ol>
  </div>
</div>
`;
